// react-github-calendar/build/index.js
'use client';
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var Calendar = require('react-activity-calendar');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);
var Calendar__default = /*#__PURE__*/_interopDefault(Calendar);

function _extends() {
  return _extends = Object.assign ? Object.assign.bind() : function (n) {
    for (var e = 1; e < arguments.length; e++) {
      var t = arguments[e];
      for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]);
    }
    return n;
  }, _extends.apply(null, arguments);
}

const transformData = (data, transformFn) => {
  if (typeof transformFn !== 'function') {
    return data;
  }
  const transformedData = transformFn(data);
  if (!Array.isArray(transformedData)) {
    throw Error(`transformData() function must return a list of Activity objects.`);
  }
  for (const d of transformedData) {
    if (!isRecord(d)) {
      throw Error(`transformData() must return a list of valid Activity objects.`);
    }
    if (typeof d.count !== 'number' || d.count < 0) {
      throw Error(`Required property "count: number" missing or invalid. Got: ${d.count}`);
    }
    if (typeof d.date !== 'string' || !/\d{4}-\d{2}-\d{2}/.test(d.date)) {
      throw Error(`Required property "date: YYYY-MM-DD" missing or invalid. Got: ${d.date}`);
    }
    if (typeof d.level !== 'number' || d.level < 0 || d.level > 4) {
      throw Error(`Required property "level: 0 | 1 | 2 | 3 | 4" missing or invalid: Got: ${d.level}.`);
    }
  }
  return transformedData;
};
const isRecord = o => Object.prototype.toString.call(o) === '[object Object]';

async function fetchCalendarData(username, year) {
  const apiUrl = 'https://github-contributions-api.jogruber.de/v4/';
  const response = await fetch(`${apiUrl}${username}?y=${year}`);
  const data = await response.json();
  if (!response.ok) {
    throw Error(`Fetching GitHub contribution data for "${username}" failed: ${data.error}`);
  }
  return data;
}
const GitHubCalendar = /*#__PURE__*/React.forwardRef(({
  username,
  year = 'last',
  labels,
  transformData: transformFn,
  transformTotalCount = true,
  throwOnError = false,
  errorMessage = `Error – Fetching GitHub contribution data for "${username}" failed.`,
  ...props
}, ref) => {
  const [data, setData] = React.useState(null);
  const [loading, setLoading] = React.useState(false);
  const [error, setError] = React.useState(null);
  const fetchData = React.useCallback(() => {
    setLoading(true);
    setError(null);
    fetchCalendarData(username, year).then(setData).catch(err => {
      if (err instanceof Error) {
        setError(err);
      }
    }).finally(() => {
      setLoading(false);
    });
  }, [username, year]);
  React.useEffect(fetchData, [fetchData]);

  // React error boundaries can't handle asynchronous code, so rethrow.
  if (error) {
    if (throwOnError) {
      throw error;
    } else {
      return /*#__PURE__*/React__default.default.createElement("div", null, errorMessage);
    }
  }
  if (loading || !data) {
    return /*#__PURE__*/React__default.default.createElement(Calendar.Skeleton, _extends({}, props, {
      loading: true
    }));
  }
  const theme = props.theme ?? gitHubTheme;
  const defaultLabels = {
    totalCount: `{{count}} contributions in ${year === 'last' ? 'the last year' : '{{year}}'}`
  };
  const totalCount = year === 'last' ? data.total['lastYear'] : data.total[year];
  return /*#__PURE__*/React__default.default.createElement(Calendar__default.default, _extends({
    data: transformData(data.contributions, transformFn),
    labels: Object.assign({}, defaultLabels, labels),
    ref: ref,
    totalCount: transformFn && transformTotalCount ? undefined : totalCount
  }, props, {
    theme: theme,
    loading: Boolean(props.loading) || loading,
    maxLevel: 4
  }));
});
GitHubCalendar.displayName = 'GitHubCalendar';
const gitHubTheme = {
  light: ['#ebedf0', '#9be9a8', '#40c463', '#30a14e', '#216e39'],
  dark: ['#161b22', '#0e4429', '#006d32', '#26a641', '#39d353']
};

exports.default = GitHubCalendar;
